From 0de9866d0a371458a37e464ca204e4c4338c808b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=D0=BD=D0=B0=D0=B1?= <nabijaczleweli@nabijaczleweli.xyz>
Date: Sun, 20 Dec 2020 14:26:24 +0100
Subject: [PATCH 1/3] Remove name checks in PPPoEDevnameHook()
X-Mutt-PGP: OS

This removes special handling of eth* and br*, instead checking any
unrecognised option for interfaceness; it's rare to see a bare "eth2" on
modern udev systems, so the requirement to prefix every interface that
could usefully exist with "nic-" is a tad odd

Based on pppd commit ccc4c4604dd9e7316fe734b41b7327aa5906de3e
("Remove name checks in the rp-pppoe plugin PPPoEDevnameHook function")
by Paul Mackerras
---
 doc/KERNEL-MODE-PPPOE |  3 +--
 src/plugin.c          | 12 ++++--------
 2 files changed, 5 insertions(+), 10 deletions(-)

diff --git a/doc/KERNEL-MODE-PPPOE b/doc/KERNEL-MODE-PPPOE
index 3dbcb4c..38cd85d 100644
--- a/doc/KERNEL-MODE-PPPOE
+++ b/doc/KERNEL-MODE-PPPOE
@@ -80,8 +80,7 @@ After that, pppoe-start should use kernel-mode PPPoE.
 
 The rp-pppoe.so plugin adds the following command-line options to pppd:
 
-ethXXX                          -- Use interface ethXXX as Ethernet interface
-brXXX                           -- Use interface brXXX as Ethernet interface
+XXX                             -- Use interface XXX as Ethernet interface
 nic-XXXX                        -- Use interface XXXX as the Ethernet interface
 
 rp_pppoe_service SERVICE_NAME   -- Specify desired service name
diff --git a/src/plugin.c b/src/plugin.c
index 902e533..d08cbac 100644
--- a/src/plugin.c
+++ b/src/plugin.c
@@ -380,16 +380,12 @@ PPPoEDevnameHook(char *cmd, char **argv, int doit)
 	return 0;
     }
 
-    /* Only do it if name is "ethXXX" or "brXXX" or what was specified
-       by rp_pppoe_dev option (ugh). */
-    /* Can also specify nic-XXXX in which case the nic- is stripped off. */
+    /* Take any otherwise-unrecognized option as a possible device name,
+       and test if it is the name of a network interface with a
+       hardware address whose sa_family is ARPHRD_ETHER.
+    */
     if (!strncmp(cmd, "nic-", 4)) {
 	cmd += 4;
-    } else {
-	if (strncmp(cmd, "eth", 3) &&
-	    strncmp(cmd, "br", 2)) {
-	    return 0;
-	}
     }
 
     /* Open a socket */
-- 
2.20.1

